src/utils/kanjiDatabase.ts

```diff
@@ -0,0 +1,230 @@
+import initSqlJs, { Database, SqlJsStatic } from 'sql.js';
+import { logger } from './logger';
+import { KANJI_BY_JLPT_QUERY, KANJI_BY_KANKEN_QUERY, KANJI_BY_JOYO_QUERY } from './sql-queries';
+import type { KanjiMetadata } from '../types/Database';
+
+const KANJI_DB_URL = 'https://github.com/migaku-official/Migaku-Kanji-Addon/blob/main/addon/kanji.db?raw=true';
+
+interface KanjiDatabaseState {
+  sql: SqlJsStatic | null;
+  db: Database | null;
+  isLoading: boolean;
+  error: string | null;
+}
+
+const kanjiDbState: KanjiDatabaseState = {
+  sql: null,
+  db: null,
+  isLoading: false,
+  error: null,
+};
+
+const filterCache = new Map<number, KanjiMetadata[]>();
+
+async function initializeSqlEngine(): Promise<SqlJsStatic | null> {
+  try {
+    if (kanjiDbState.sql) {
+      logger.debug('Using existing SQL.js instance for kanji DB');
+      return kanjiDbState.sql;
+    }
+
+    logger.debug('Initializing SQL.js for kanji DB...');
+    
+    const SQL = await initSqlJs({
+      locateFile: (file: string) => {
+        logger.debug(`Locating file: ${file}`);
+        if (file.endsWith('.wasm')) {
+          return 'https://cdn.jsdelivr.net/npm/sql.js@1.13.0/dist/sql-wasm.wasm';
+        }
+        return file;
+      },
+    });
+    
+    if (!SQL) {
+      throw new Error('SQL.js initialization returned null');
+    }
+    
+    logger.debug('SQL.js initialized successfully for kanji DB');
+    kanjiDbState.sql = SQL;
+    return SQL;
+  } catch (err) {
+    logger.error('Failed to initialize SQL.js for kanji DB:', err);
+    return null;
+  }
+}
+
+async function loadKanjiDatabase(): Promise<Database | null> {
+  try {
+    if (kanjiDbState.db) {
+      logger.debug('Using existing kanji database instance');
+      return kanjiDbState.db;
+    }
+
+    if (kanjiDbState.isLoading) {
+      logger.debug('Kanji database already loading, waiting...');
+      let attempts = 0;
+      while (kanjiDbState.isLoading && attempts < 50) {
+        await new Promise((resolve) => setTimeout(resolve, 100));
+        attempts++;
+      }
+      return kanjiDbState.db;
+    }
+
+    kanjiDbState.isLoading = true;
+    kanjiDbState.error = null;
+
+    logger.debug('Fetching kanji.db from GitHub...');
+    
+    const arrayBuffer = await new Promise<ArrayBuffer>((resolve, reject) => {
+      // @ts-ignore
+      GM.xmlHttpRequest({
```

Comment:
```suggestion
      GM_xmlhttpRequest({
```
We should use `GM_xmlhttpRequest`, importing it like:
```
// @ts-ignore
import { GM_xmlhttpRequest } from 'monkey';
```